### PR TITLE
Reject future blocks

### PR DESCRIPTION
**Pull request by teor2345**
Reject blocks that are more than 2 hours in the future, according to the node's clock.

~~I'll be able to complete this PR, once we've finished the async `AddBlock()` changes in PR #491.~~

Edit: PR #491 wasn't actually needed.

### CODE REVIEW

**Comment by teor2345** [comment]
In this case, the time check is a quick check, so we can just do it in `call()`, and if it fails, return `async { Err() }` as the Future.

But we'll still need PR #491 to do expensive checks in the async block.

**Review by hdevalence — DISMISSED** [review]
:+1:

File: zebra-consensus/src/verify.rs
Comment (dconnolly) [inline zebra-consensus/src/verify.rs]: ☝️

File: zebra-consensus/src/verify.rs
Comment (dconnolly) [inline zebra-consensus/src/verify.rs]: 👍👍👍

**Review by dconnolly — DISMISSED** [review]
Writing a test to cover this seems annoying to set up so I'm approving without a test, but we may want to create an issue to make one and make it open to contributors / good-first-issue.

**Comment by teor2345** [comment]
> Writing a test to cover this seems annoying to set up so I'm approving without a test, but we may want to create an issue to make one and make it open to contributors / good-first-issue.

I think a test for "2 hours in the future" should be reasonably easy, let me see if I can set one up.

It will be harder to test overflow, I might need to refactor to do that.

**Comment by teor2345** [comment]
I rebased this branch to handle the conflicts with the comments I cherry-picked from PR #491.

> > Writing a test to cover this seems annoying to set up so I'm approving without a test, but we may want to create an issue to make one and make it open to contributors / good-first-issue.
> 
> I think a test for "2 hours in the future" should be reasonably easy, let me see if I can set one up.
> 
> It will be harder to test overflow, I might need to refactor to do that.

I tested:
* 2 hours in the future,
* 2 hours and 1 second in the future,
* a bunch of nearby times,
* an old block, and
* a bunch of significant time constants.

`Utc.timestamp()` panics when the timestamp is out of range. Its valid range is smaller than i64, so I don't think overflow is possible. (But if it is, the check handles it correctly, by returning an error.)

The `Utc.timestamp()` valid range is larger than the Zcash time range (u32), so all Zcash times are valid in Rust. I've made a note in #477 to check for overflows when adding times from multiple blocks.

**Comment by teor2345** [comment]
I've set this PR to draft, because I need to fix an issue with early state commits, as discussed here:
https://github.com/ZcashFoundation/zebra/pull/502#discussion_r442125868

I should also add some tests, to make sure that the state isn't changed on verification errors.

File: zebra-consensus/src/verify.rs
Comment (teor2345) [inline zebra-consensus/src/verify.rs]: As discussed in #502, we can't conditionally await an AddBlock future. Because the state service can commit changes in `call()`, or in the Future's async block (that is, in await).

File: zebra-chain/src/block.rs
Comment (yaahc) [inline zebra-chain/src/block.rs:151]: If we're worried about truncating casts we could alternatively use `try_from().unwrap()` to just panic on any overflow

File: zebra-consensus/src/verify/block.rs
Comment (yaahc) [inline zebra-consensus/src/verify/block.rs]: TIOLI: I don't think this benefits from being broken out into a separate function, I'd prefer to start the fn with

```rust
let block_header_time = block.header.time;
let now = Utc::now();

// The rest of node_time_check_helper's function body...
```

File: zebra-chain/src/block.rs
Comment (teor2345) [reply zebra-chain/src/block.rs:151]: I don't think truncation is possible until we start generating our own timestamps. Even then, our block validation code should detect any truncation, because the time won't be in the valid range (within 90 minutes of the previous blocks' median, and less 2 hours into the future according to the node's clock).

I'm hesitant about panics in general - if they can be triggered remotely, they can lead to a denial of service attack.

File: zebra-consensus/src/verify/block.rs
Comment (teor2345) [reply zebra-consensus/src/verify/block.rs]: You're right, it made more sense before I split out the node_time_check_helper() for testing.

Comment (teor2345) [reply zebra-consensus/src/verify/block.rs]: Fixed in 347f5ba7b494795fc2f9801b9b190238f27fd88f.

File: zebra-chain/src/block.rs
Comment (teor2345) [reply zebra-chain/src/block.rs:151]: I update the comment to talk about the block time checks in 153a96848a612b74240e7c511e3edf199dfeae7a.

File: zebra-consensus/src/verify.rs
Comment (teor2345) [reply zebra-consensus/src/verify.rs]: Fixed in 589f1cebffd8ef1d0e7599cdc0445f5aaec1397f.

**Comment by teor2345** [comment]
I think we can let GitHub squash and merge this PR.

(I could do a manual squash into a time check commit, a BlockVerifier refactor commit, and other changes. Happy to do that if people think it's worth the effort.)

File: zebra-consensus/src/verify.rs
Comment (dconnolly) [inline zebra-consensus/src/verify.rs:305]: 👍

File: zebra-consensus/src/verify.rs
Comment (dconnolly) [inline zebra-consensus/src/verify.rs:297]: Yay! Thanks for covering this case

**Review by dconnolly — DISMISSED** [review]
All these tests fill me with joy ☺️

File: zebra-consensus/src/verify.rs
Comment (yaahc) [inline zebra-consensus/src/verify.rs:270]: I wouldn't default to doing string comparisons, if there's ever a specific error that we need to check for that is being constructed from `str -> Box<dyn Error>` we should instead change the error to be a proper error type that we can downcast to, using string's content as a proxy for a typed error is a recipe for disaster.

File: zebra-consensus/src/verify.rs
Comment (yaahc) [inline zebra-consensus/src/verify.rs:320]: I swear I'm going to kill all of these `map_err`s and sooner rather than later.

 _shakes fist at annoying conversion logic necessitated by tower_

**Review by hdevalence — DISMISSED** [review]
TIOLI: I think the `BlockVerifier` belongs in the `block` submodule, and generally I think we should keep all of the contents of the code in the leaves of the module tree and then re-export as necessary.

File: zebra-consensus/src/verify.rs
Comment (teor2345) [reply zebra-consensus/src/verify.rs:270]: I'll fix these comments, possibly in a future PR.

**Comment by dconnolly** [comment]
Looks like this needs a rebase but LGTM! 💃

**Comment by teor2345** [comment]
I rebased and made the following changes:
* zebra-test module name change
* color-eyre/eyre module import changes

I'll merge after CI passes, I don't think the differences are large enough for a review.

**Comment by teor2345** [comment]
The checks passed, and the changes are minor, so I merged this PR.

**Comment by teor2345** [comment]
> TIOLI: I think the `BlockVerifier` belongs in the `block` submodule, and generally I think we should keep all of the contents of the code in the leaves of the module tree and then re-export as necessary.

Fixed in #538 in e2999fc958a172e0f296e5b439d4a6a3d0e1a80a.

File: zebra-consensus/src/verify.rs
Comment (teor2345) [reply zebra-consensus/src/verify.rs:270]: Fixed in #538 in 5f96b3239843790f498cbcdd45ad556db949b6ae for verify/block

Comment (teor2345) [reply zebra-consensus/src/verify.rs:270]: Fixed in #538 in 609de233128140a92f4866c67f847c24f4dd192c for checkpoint

**Pull request merged**
